Improve editor gui

andreasb242 opened the pull request:
* New file menu and toolbar, configured by a property file, same actions for Toolbar and Menu
   * Keyboard shortcuts also defined in a seperate property file
```INI
# Main menubar
menubar=file,edit,view,help

## Translations:
## Key menu.xxx is used as translation key
## "---" is a separator

menu.file=new,open,save,save-as,open-map-folder,---,play,---,export-image,---,quit
menu.edit=undo,redo
menu.view=zoom-in,zoom-out,zoom100,---,show-statistic,show-map-settings,show-tools,---,show-errors,goto-error
menu.help=manual,about

## Toolbar
toolbar=new,open,save,---,undo,redo,---,play,---,zoom-in,zoom-out,zoom100,---,show-statistic,show-map-settings,---,show-errors,goto-error,---,player-spinner
```
* Sidebar with Tabs for errors and tools, tools have icons (not all yet)
![s3-4](https://cloud.githubusercontent.com/assets/822958/12064712/a73f768c-afca-11e5-878f-b04e3ab0dabc.png)
   * Sidebar is resizeable
![s3-5](https://cloud.githubusercontent.com/assets/822958/12064714/a7444996-afca-11e5-9445-d04e3ca3ae7b.png)
* Open or create a new file from editor, start a new instance
* New start Dialog, open mask list formatted
   * Additional Last used tab added, which shows (max.) 10 last used maps
![s3-1](https://cloud.githubusercontent.com/assets/822958/12064710/a73ae374-afca-11e5-8f55-5943fe4b1025.png)
![s3-2](https://cloud.githubusercontent.com/assets/822958/12064711/a73eaef0-afca-11e5-9ad7-bb2455afb80d.png)
![s3-3](https://cloud.githubusercontent.com/assets/822958/12064713/a74030cc-afca-11e5-94d5-d766e844e79f.png)

andreasb242 commented:
"Save as" and *Export as image" is not yet implemented.

andreasb242 commented:
Preview is now displayed in the Open window... I coulden't find an example, and the documentation could be better ;-)
![s3-10](https://cloud.githubusercontent.com/assets/822958/12070966/f47d6684-b092-11e5-921c-4e6ffa637fb0.png)
Landscape type is now translated, and the color is displayed, as in the tool tree also
![s3-11](https://cloud.githubusercontent.com/assets/822958/12070967/05e22c5c-b093-11e5-90f6-4f5898295239.png)
Player selection now shows the player color, and the height tools have icons.
![s3-12](https://cloud.githubusercontent.com/assets/822958/12070969/19e69698-b093-11e5-90a2-f7ff05dbdfc4.png)

And a lot of small changes...

michaelzangl commented on jsettlers.mapcreator/src/jsettlers/mapcreator/main/DataTester.java:
```diff
@@ -49,21 +52,48 @@
+			notifyAll();
+		}
+		try {
+			thread.join(100);
```
Why the timeout? We should at least handle the case that the thread runs into a timeout.

andreasb242 replied:
Currently the editor is always quit with System.exit, because the re are a lot of threads which aren't stopped.
But you're right... It makes no sense as it is currently, i add a thread.stop() to kill after, OK?
```java
			thread.join(100);
			thread.stop();
```
Not the nice solution, but its currently anyway only at the end, and the thread should have enough time to quit.

michaelzangl replied:
thread.stop() is very bad. You could throw an exception if the join was not successful. Or simply remove the timeout: This allows users to file a bug report if the thread hangs and we can trace the source of the problem easier.

andreasb242 replied:
System.exit is also very bad...
This is not my code. I only add the dipose().

For the moment it's working.
Can we do later a refactoring, and use an ExecutorService service instead?
Then we don't have the problem. 
Then also the whole synchronisation stuff can be removed, if there a new check should be done, we create a new job, which is executed in the Threadpool (single threaded).

OK for you?

Off topic, but I hope you read this here:
@michaelzangl you already have 144 commits on this project, you may know it:
Was there already a discussion about error loggin and exception handling?

I added a default thread exception handler in the main of the Map Editor, which catches all unhandled exception, display a popup and print the stacktrace to stderr.

I also used this popup on all important catch blocks I found, before only the exception was logged, you clocked somewhere and nothing happend. Really annoying as user.

I would generally show an error if an exception happens which matters.

And I would prefer to use SLF4J, you have a lot more possibilities (turn off for performance, detailed logging to stdout on debugging etc.)

I didn't see any discussion about this topic...

michaelzangl replied:
I have more than 1k commits, github is not counting all of them.

The DataTester was simply killed on exit because we know that we do not need it's data then. Stopping it by force may lead to corrupt DataTester data. I cannot guarantee that the rest of the code will work then.

Yes, there were thoughts about error logging but for the moment we just stuck to System.err.

We just don't have anything like a clear GUI exception handling. The GUI always attempts to catch them (if any could happen) or it simply crashes - in hope that users report the bugs they find.

Adding a good logging system would be a big improvement to the Project. #259

andreasb242 replied:
The Thread is not the only problem, the List with the errors, is handed over in the UI Thread, but not copied, so the The Checker thread and the UI Thread are working on the same list... Not good, I have to fix this...

andreasb242 replied:
I didn't see your commend, github didn't automatically refresh...
I'll fix the Datatester, and see if I can improve it a little bit with the new sidebar.
I'll extends my swing exception handler in the Map Editor. It can be later moved to use also for the game. Now I'm only working on the Map Editor, I don't know the Code enough good.
(And the Map Editor don't have to run on Android, makes it for the beginning a lot easyer...)

Logging is then may the next task...

michaelzangl commented:
Well done. Yes, the editor was a bit hacked. This makes the editor much more usable.

andreasb242 commented:
Thank you :-)
Only the shape type selection is not good, really bad usability...
I'll have to think about, how a good selection can look like, currently checkboxes and settings appear / disappear on selection tools...

michaelzangl commented:
Can you merge this branch? I did some enhancements on the error reporting.

https://github.com/jsettlers/settlers-remake/tree/andreasb242-improve-editor-gui

andreasb242 commented:
Should be merged, but the automatically build is not starting any more?

andreasb242 commented:
My local build is working. I changed the ant file, because I added JUnitTests for the Map Editor. Can somebody check whats wrong? I think all should run... :-(

andreasb242 commented:
Build is working, but with disabled JUnitTest

@michaelzangl Can you have a look, the File "ValidationTest.java_" contains the JUnitTest for the MapCreator, which has to be added to the Ant Build path.
I couldn't get it build with ant...

andreasb242 commented:
The most problems which are detected in the editor can also be automatically fixed.
The original Settler 3 editor blocks painting near the boards, not my prefered way, I like it much more like Eclipse: Let the user do all mistakes, but give hime the option to automatically fix it.

Only the first simple implementation is done, to take a screenshot :-)

![s3-30](https://cloud.githubusercontent.com/assets/822958/12075031/be978528-b172-11e5-82fd-81a67befe9d6.png)
![s3-31](https://cloud.githubusercontent.com/assets/822958/12075033/c095e55e-b172-11e5-889d-27be86a0ee1c.png)
![s3-32](https://cloud.githubusercontent.com/assets/822958/12075034/c1e6ebec-b172-11e5-9166-c6c4c6b08d1c.png)

andreasb242 commented:
Added icons for Shape types, display all types always, but disable the not supported types, makes the UI less flickering.

@michaelzangl 
I need some description / help: I would expect that I can draw a straight line with the line tool, but I couldn't figure out how?

![s3-40](https://cloud.githubusercontent.com/assets/822958/12075816/d6ac04c0-b18d-11e5-94e3-c4ca67d35819.png)

andreasb242 commented:
Again some screenshots of my progress, and some description...

Open map with correct preview display, search within map name, description and ID
![s3-50](https://cloud.githubusercontent.com/assets/822958/12096925/06c3169e-b319-11e5-8fff-1e33e0ae5e48.png)

Menu option to display resources always. If you select a resource tool or a tool to build a mine or a fishing house, the resources are also displayed (even if the checkbox is not checked)
![s3-51](https://cloud.githubusercontent.com/assets/822958/12096946/167d3c36-b319-11e5-9376-bb2d03a4916f.png)

You can do a lot of non working stuff, e.g. if you draw gras over water with fishes, the fishes stay there.
You can place settlers out of the land of the owner... (you may would like to build a tower later?)
The errors are displayed at the right sidebar, errors are grouped, 10 settler which are next together will displayed as one error, instead of one error per settler...
![s3-52](https://cloud.githubusercontent.com/assets/822958/12096970/40c98dfa-b319-11e5-8263-e6ec3e93f93c.png)

Now there is an auto fix option in the new "Error / Warning" menu, or you can press F9...
![s3-54](https://cloud.githubusercontent.com/assets/822958/12097023/889ec1cc-b319-11e5-9508-593ab99de408.png)

And the errors are gone!
Not all errors can be fixed automatically, and in this case the fishes and settlers are simple deleted...
![s3-55](https://cloud.githubusercontent.com/assets/822958/12097036/a0d84f10-b319-11e5-9307-0eaf82d7dedd.png)

andreas-eberle commented:
I just opened a new PR #272 as follow up of this PR. Therefore I will close this PR. All work will be continued in the new PR.
